### PR TITLE
utils: removed deprecated check and function for OpenSSL compatiblity

### PR DESCRIPTION
**Pull request by azavalla**
refs https://github.com/bitcoin/bitcoin/issues/11601

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
It looks like openssl removed the threading stuff out from under us - we should proabbly do something like check OPENSSL_THREADS (ie if openssl was built with threading support - can we also do this at runtime?) and fail to build/run if it wasn't, only setting the locking callbacks if the openssl version is low enough that they are needed.

**Comment by laanwj** [comment]
> It looks like openssl removed the threading stuff out from under us

What I don't understand is why this doesn't fail because it still calls `CRYPTO_set_locking_callback`?

> can we also do this at runtime?

Maybe - though even checking the OpenSSL version at runtime isn't straightforward. E.g. there used to be a function `SSLeay` to get a version number which they suddenly removed and replaced with `OpenSSL_version_num` (see https://github.com/openssl/openssl/issues/862). Also would have to consider alternative implementations such as LibreSSL and PolarSSL, which try to be compatible but with whoknowswhat...

What a mess.

**Comment by TheBlueMatt** [comment]
> What I don't understand is why this doesn't fail because it still calls CRYPTO_set_locking_callback?

The functions themselves were kept, except they are now just #defines to nothing, the constant was removed.

I am somewhat concerned that the current code may break if OpenSSL is complied without threading support, though I'd hope that's rather rare in any case.

**Comment by laanwj** [comment]
> I am somewhat concerned that the current code may break if OpenSSL is complied without threading support, though I'd hope that's rather rare in any case.

I had hoped that the thread-related functions would be missing from the library in that case, causing either compilation error or a runtime error in the dynamic linker. 
(one'd hope they don't #define them away in the case of no thread support, only because the API for thread support doesn't need them anymore, but who knows..)

**Comment by laanwj** [comment]
In any case this can't be merged as-is. It breaks the tests, and breaks compatibility with older OpenSSL. At the least it needs `#ifdef` to delimit the OpenSSL versions.

**Comment by azavalla** [comment]
They removed the thread-related functions and then `#define`d them again as `no-ops`.
This means that:
- Everything will work just fine for a openssl version low enough.
- Compilation will fail for the openssl versions where they removed those functions.
- Possibly obscure race conditions at run time for openssl versions with the `no-op` defines?

Is that correct?

File: src/util.cpp [hunk 1]
```diff
@@ -107,7 +107,7 @@ std::atomic<uint32_t> logCategories(0);
 static std::unique_ptr<CCriticalSection[]> ppmutexOpenSSL;
 void locking_callback(int mode, int i, const char* file, int line) NO_THREAD_SAFETY_ANALYSIS
 {
-    if (mode & CRYPTO_LOCK) {
+    if (mode) {
```
Comment (laanwj) [inline src/util.cpp]: Does the removal of CRYPTO_LOCK here have any influence on older versions of OpenSSL which need this?
Might be better to add the define here, if it's not defined.

File: src/util.cpp [hunk 2]
```diff
@@ -141,8 +141,6 @@ class CInit
     }
     ~CInit()
     {
-        // Securely erase the memory used by the PRNG
-        RAND_cleanup();
```
Comment (laanwj) [inline src/util.cpp]: We'd still want to call this function for versions of OpenSSL that need it.

**Comment by laanwj** [comment]
Ping @azavalla

**Comment by laanwj** [comment]
Closing due to inactivity, added "up for grabs" label, in case anyone else wants to have a go at this.
(or let me know when you want to pick this back up)

**Comment by adamjonas** [comment]
Can remove up for grabs label now that #11601 is closed.

**Pull request closed without merging**